File: app/fsharp-2/maui/tests/PuyoPuyo.Tests/Domain/BoardTests.fs
module PuyoPuyo.Tests.Domain.BoardTests

open Xunit
open FsUnit.Xunit
open PuyoPuyo.Core.Domain

[<Fact>]
let ``空のボードを作成できる`` () =
    // Arrange & Act
    let board = Board.create 6 13

    // Assert
    board.Cols |> should equal 6
    board.Rows |> should equal 13

[<Fact>]
let ``作成直後のボードはすべて空である`` () =
    // Arrange & Act
    let board = Board.create 6 13

    // Assert
    for y in 0 .. board.Rows - 1 do
        for x in 0 .. board.Cols - 1 do
            Board.getCell board x y |> should equal Cell.Empty

[<Fact>]
let ``ボードにぷよを配置できる`` () =
    // Arrange
    let board = Board.create 6 13

    // Act
    let newBoard = Board.setCell board 2 10 (Cell.Filled PuyoColor.Red)

    // Assert
    Board.getCell newBoard 2 10 |> should equal (Cell.Filled PuyoColor.Red)

[<Fact>]
let ``ボードにぷよを配置しても元のボードは変更されない`` () =
    // Arrange
    let board = Board.create 6 13

    // Act
    let newBoard = Board.setCell board 2 10 (Cell.Filled PuyoColor.Red)

    // Assert
    Board.getCell board 2 10 |> should equal Cell.Empty
    Board.getCell newBoard 2 10 |> should equal (Cell.Filled PuyoColor.Red)

[<Fact>]
let ``ぷよペアをボードに固定できる`` () =
    // Arrange
    let board = Board.create 6 13
    let pair = PuyoPair.create 3 10 PuyoColor.Red PuyoColor.Green 0

    // Act
    let newBoard = Board.fixPuyoPair board pair

    // Assert
    let (pos1, pos2) = PuyoPair.getPositions pair
    let (x1, y1) = pos1
    let (x2, y2) = pos2
    Board.getCell newBoard x1 y1 |> should equal (Cell.Filled PuyoColor.Red)
    Board.getCell newBoard x2 y2 |> should equal (Cell.Filled PuyoColor.Green)

[<Fact>]
let ``ぷよペアを固定しても元のボードは変更されない`` () =
    // Arrange
    let board = Board.create 6 13
    let pair = PuyoPair.create 3 10 PuyoColor.Red PuyoColor.Green 0

    // Act
    let newBoard = Board.fixPuyoPair board pair

    // Assert
    let (pos1, pos2) = PuyoPair.getPositions pair
    let (x1, y1) = pos1
    Board.getCell board x1 y1 |> should equal Cell.Empty // 元のボードは空のまま
    Board.getCell newBoard x1 y1 |> should equal (Cell.Filled PuyoColor.Red) // 新しいボードには固定

[<Fact>]
let ``横に4つ並んだぷよを検出できる`` () =
    // Arrange
    let board = Board.create 6 13

    let board =
        board
        |> fun b -> Board.setCell b 0 12 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 1 12 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 2 12 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 3 12 (Cell.Filled PuyoColor.Red)

    // Act
    let groups = Board.findConnectedGroups board

    // Assert
    groups |> List.length |> should equal 1
    groups |> List.head |> List.length |> should equal 4

[<Fact>]
let ``縦に4つ並んだぷよを検出できる`` () =
    // Arrange
    let board = Board.create 6 13

    let board =
        board
        |> fun b -> Board.setCell b 2 9 (Cell.Filled PuyoColor.Green)
        |> fun b -> Board.setCell b 2 10 (Cell.Filled PuyoColor.Green)
        |> fun b -> Board.setCell b 2 11 (Cell.Filled PuyoColor.Green)
        |> fun b -> Board.setCell b 2 12 (Cell.Filled PuyoColor.Green)

    // Act
    let groups = Board.findConnectedGroups board

    // Assert
    groups |> List.length |> should equal 1
    groups |> List.head |> List.length |> should equal 4

[<Fact>]
let ``L字型につながった5つのぷよを検出できる`` () =
    // Arrange
    let board = Board.create 6 13

    let board =
        board
        |> fun b -> Board.setCell b 1 10 (Cell.Filled PuyoColor.Blue)
        |> fun b -> Board.setCell b 1 11 (Cell.Filled PuyoColor.Blue)
        |> fun b -> Board.setCell b 1 12 (Cell.Filled PuyoColor.Blue)
        |> fun b -> Board.setCell b 2 12 (Cell.Filled PuyoColor.Blue)
        |> fun b -> Board.setCell b 3 12 (Cell.Filled PuyoColor.Blue)

    // Act
    let groups = Board.findConnectedGroups board

    // Assert
    groups |> List.length |> should equal 1
    groups |> List.head |> List.length |> should equal 5

[<Fact>]
let ``3つ以下のぷよは検出されない`` () =
    // Arrange
    let board = Board.create 6 13

    let board =
        board
        |> fun b -> Board.setCell b 0 12 (Cell.Filled PuyoColor.Yellow)
        |> fun b -> Board.setCell b 1 12 (Cell.Filled PuyoColor.Yellow)
        |> fun b -> Board.setCell b 2 12 (Cell.Filled PuyoColor.Yellow)

    // Act
    let groups = Board.findConnectedGroups board

    // Assert
    groups |> List.length |> should equal 0

[<Fact>]
let ``指定した位置のぷよを消去できる`` () =
    // Arrange
    let board = Board.create 6 13

    let board =
        board
        |> fun b -> Board.setCell b 0 12 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 1 12 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 2 12 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 3 12 (Cell.Filled PuyoColor.Red)

    // Act
    let positions = [ (0, 12); (1, 12); (2, 12); (3, 12) ]
    let newBoard = board |> Board.clearPuyos positions

    // Assert
    Board.getCell newBoard 0 12 |> should equal Cell.Empty
    Board.getCell newBoard 1 12 |> should equal Cell.Empty
    Board.getCell newBoard 2 12 |> should equal Cell.Empty
    Board.getCell newBoard 3 12 |> should equal Cell.Empty

[<Fact>]
let ``重力を適用すると浮いているぷよが落ちる`` () =
    // Arrange
    let board = Board.create 6 13

    let board =
        board
        |> fun b -> Board.setCell b 2 8 (Cell.Filled PuyoColor.Green) // 浮いているぷよ
        |> fun b -> Board.setCell b 2 12 (Cell.Filled PuyoColor.Red) // 下にあるぷよ

    // Act
    let newBoard = Board.applyGravity board

    // Assert
    Board.getCell newBoard 2 8 |> should equal Cell.Empty
    Board.getCell newBoard 2 11 |> should equal (Cell.Filled PuyoColor.Green) // 落ちた
    Board.getCell newBoard 2 12 |> should equal (Cell.Filled PuyoColor.Red)

[<Fact>]
let ``重力を適用すると複数のぷよが落ちる`` () =
    // Arrange
    let board = Board.create 6 13

    let board =
        board
        |> fun b -> Board.setCell b 1 5 (Cell.Filled PuyoColor.Blue)
        |> fun b -> Board.setCell b 1 6 (Cell.Filled PuyoColor.Yellow)
        |> fun b -> Board.setCell b 1 12 (Cell.Filled PuyoColor.Red)

    // Act
    let newBoard = Board.applyGravity board

    // Assert
    Board.getCell newBoard 1 5 |> should equal Cell.Empty
    Board.getCell newBoard 1 6 |> should equal Cell.Empty
    Board.getCell newBoard 1 10 |> should equal (Cell.Filled PuyoColor.Blue)
    Board.getCell newBoard 1 11 |> should equal (Cell.Filled PuyoColor.Yellow)
    Board.getCell newBoard 1 12 |> should equal (Cell.Filled PuyoColor.Red)

[<Fact>]
let ``連鎖が発生する_基本ケース`` () =
    // Arrange: 赤ぷよ2x2と青ぷよ縦3+横1を配置
    // 赤ぷよが消えると青ぷよが落下して4つつながる
    let board = Board.create 6 13

    let board =
        board
        |> fun b -> Board.setCell b 1 11 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 2 11 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 1 12 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 2 12 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 3 11 (Cell.Filled PuyoColor.Blue)
        |> fun b -> Board.setCell b 2 8 (Cell.Filled PuyoColor.Blue)
        |> fun b -> Board.setCell b 2 9 (Cell.Filled PuyoColor.Blue)
        |> fun b -> Board.setCell b 2 10 (Cell.Filled PuyoColor.Blue)

    // Act: 再帰的に消去と重力を適用
    let (finalBoard, isZenkeshi) = Board.clearAndApplyGravityRepeatedly board

    // Assert: 全て消えているはず（2連鎖発生）
    Board.getCell finalBoard 1 11 |> should equal Cell.Empty
    Board.getCell finalBoard 2 11 |> should equal Cell.Empty
    Board.getCell finalBoard 3 11 |> should equal Cell.Empty
    // 全消しになっている
    isZenkeshi |> should equal true

[<Fact>]
let ``連鎖が発生しない_消去パターンなし`` () =
    // Arrange: 消去パターンがない状態
    let board = Board.create 6 13

    let board =
        board
        |> fun b -> Board.setCell b 1 12 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 2 12 (Cell.Filled PuyoColor.Blue)
        |> fun b -> Board.setCell b 3 12 (Cell.Filled PuyoColor.Green)

    // Act
    let (finalBoard, isZenkeshi) = Board.clearAndApplyGravityRepeatedly board

    // Assert: 何も変わっていないはず
    Board.getCell finalBoard 1 12 |> should equal (Cell.Filled PuyoColor.Red)
    Board.getCell finalBoard 2 12 |> should equal (Cell.Filled PuyoColor.Blue)
    Board.getCell finalBoard 3 12 |> should equal (Cell.Filled PuyoColor.Green)
    // 全消しにならない
    isZenkeshi |> should equal false

[<Fact>]
let ``盤面上のぷよがすべて消えると全消しになる`` () =
    // Arrange: 4つの赤ぷよを配置
    let board = Board.create 6 13

    let board =
        board
        |> fun b -> Board.setCell b 1 10 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 2 10 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 1 11 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 2 11 (Cell.Filled PuyoColor.Red)

    // Act: 消去判定と実行
    let groups = Board.findConnectedGroups board
    let positions = groups |> List.concat
    let clearedBoard = Board.clearPuyos positions board

    // 全消し判定
    let isZenkeshi = Board.checkZenkeshi clearedBoard

    // Assert: 全消しになっていることを確認
    isZenkeshi |> should equal true

[<Fact>]
let ``盤面上にぷよが残っていると全消しにならない`` () =
    // Arrange: 赤ぷよ4つと青ぷよ1つを配置
    let board = Board.create 6 13

    let board =
        board
        |> fun b -> Board.setCell b 1 10 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 2 10 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 1 11 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 2 11 (Cell.Filled PuyoColor.Red)
        |> fun b -> Board.setCell b 3 11 (Cell.Filled PuyoColor.Blue)

    // Act: 消去判定と実行（赤ぷよのみ消える）
    let groups = Board.findConnectedGroups board
    let positions = groups |> List.concat
    let clearedBoard = Board.clearPuyos positions board

    // 全消し判定
    let isZenkeshi = Board.checkZenkeshi clearedBoard

    // Assert: 全消しにならない（青ぷよが残る）
    isZenkeshi |> should equal false

[<Fact>]
let ``消去パターンがない場合でも着地後は重力が適用される`` () =
    // Arrange: 空中に浮いたぷよと下にぷよがある状態
    let board = Board.create 6 13

    let board =
        board
        |> fun b -> Board.setCell b 2 8 (Cell.Filled PuyoColor.Red) // 空中に浮いている
        |> fun b -> Board.setCell b 2 12 (Cell.Filled PuyoColor.Blue) // 下にある
        |> fun b -> Board.setCell b 3 12 (Cell.Filled PuyoColor.Green) // 下にある

    // Act: 消去パターンがないが、連鎖処理を実行（重力が適用されるべき）
    let (finalBoard, isZenkeshi) = Board.clearAndApplyGravityRepeatedly board

    // Assert: 赤ぷよが落下している
    Board.getCell finalBoard 2 8 |> should equal Cell.Empty // 元の位置は空
    Board.getCell finalBoard 2 11 |> should equal (Cell.Filled PuyoColor.Red) // 落下した位置
    Board.getCell finalBoard 2 12 |> should equal (Cell.Filled PuyoColor.Blue) // 元のまま
    Board.getCell finalBoard 3 12 |> should equal (Cell.Filled PuyoColor.Green) // 元のまま

File: app/csharp-2/maui/PuyoPuyoMAUI.Tests/Domain/BoardTests.cs
using FluentAssertions;
using PuyoPuyoMAUI.Domain;
using Xunit;

namespace PuyoPuyoMAUI.Tests.Domain;

public class BoardTests
{
    [Fact]
    public void 空のボードを作成できる()
    {
        // Arrange & Act
        var board = Board.Create(6, 13);

        // Assert
        board.Cols.Should().Be(6);
        board.Rows.Should().Be(13);
    }

    [Fact]
    public void 作成直後のボードはすべて空である()
    {
        // Arrange & Act
        var board = Board.Create(6, 13);

        // Assert
        for (int y = 0; y < board.Rows; y++)
        {
            for (int x = 0; x < board.Cols; x++)
            {
                board.GetCell(x, y).Should().Be(Cell.Empty);
            }
        }
    }

    [Fact]
    public void ボードにぷよを配置できる()
    {
        // Arrange
        var board = Board.Create(6, 13);

        // Act
        var newBoard = board.SetCell(2, 10, Cell.Filled(PuyoColor.Red));

        // Assert
        newBoard.GetCell(2, 10).Should().Be(Cell.Filled(PuyoColor.Red));
    }

    [Fact]
    public void ボードにぷよを配置しても元のボードは変更されない()
    {
        // Arrange
        var board = Board.Create(6, 13);

        // Act
        var newBoard = board.SetCell(2, 10, Cell.Filled(PuyoColor.Red));

        // Assert
        board.GetCell(2, 10).Should().Be(Cell.Empty);
        newBoard.GetCell(2, 10).Should().Be(Cell.Filled(PuyoColor.Red));
    }

    [Fact]
    public void 範囲外のセルを取得すると空のセルが返される()
    {
        // Arrange
        var board = Board.Create(6, 13);

        // Act & Assert
        board.GetCell(-1, 0).Should().Be(Cell.Empty);
        board.GetCell(0, -1).Should().Be(Cell.Empty);
        board.GetCell(6, 0).Should().Be(Cell.Empty);
        board.GetCell(0, 13).Should().Be(Cell.Empty);
    }

    [Fact]
    public void ぷよペアをボードに固定できる()
    {
        // Arrange
        var board = Board.Create(6, 13);
        var pair = new PuyoPair(3, 11, PuyoColor.Red, PuyoColor.Green, 0);

        // Act
        var newBoard = board.FixPuyoPair(pair);

        // Assert
        newBoard.GetCell(3, 11).Should().Be(Cell.Filled(PuyoColor.Red));
        newBoard.GetCell(3, 10).Should().Be(Cell.Filled(PuyoColor.Green));
    }

    [Fact]
    public void ぷよペアを固定しても元のボードは変更されない()
    {
        // Arrange
        var board = Board.Create(6, 13);
        var pair = new PuyoPair(3, 11, PuyoColor.Red, PuyoColor.Green, 0);

        // Act
        var newBoard = board.FixPuyoPair(pair);

        // Assert
        board.GetCell(3, 11).Should().Be(Cell.Empty);
        board.GetCell(3, 10).Should().Be(Cell.Empty);
        newBoard.GetCell(3, 11).Should().Be(Cell.Filled(PuyoColor.Red));
        newBoard.GetCell(3, 10).Should().Be(Cell.Filled(PuyoColor.Green));
    }

    [Fact]
    public void FindConnectedGroups_横一列の4つのぷよを検出できる()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(0, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(3, 12, Cell.Filled(PuyoColor.Red));

        // Act
        var groups = board.FindConnectedGroups();

        // Assert
        groups.Should().HaveCount(1);
        groups[0].Should().HaveCount(4);
    }

    [Fact]
    public void FindConnectedGroups_縦一列の4つのぷよを検出できる()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(2, 9, Cell.Filled(PuyoColor.Green))
            .SetCell(2, 10, Cell.Filled(PuyoColor.Green))
            .SetCell(2, 11, Cell.Filled(PuyoColor.Green))
            .SetCell(2, 12, Cell.Filled(PuyoColor.Green));

        // Act
        var groups = board.FindConnectedGroups();

        // Assert
        groups.Should().HaveCount(1);
        groups[0].Should().HaveCount(4);
    }

    [Fact]
    public void FindConnectedGroups_L字型の5つのぷよを検出できる()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(1, 10, Cell.Filled(PuyoColor.Blue))
            .SetCell(1, 11, Cell.Filled(PuyoColor.Blue))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Blue))
            .SetCell(2, 12, Cell.Filled(PuyoColor.Blue))
            .SetCell(3, 12, Cell.Filled(PuyoColor.Blue));

        // Act
        var groups = board.FindConnectedGroups();

        // Assert
        groups.Should().HaveCount(1);
        groups[0].Should().HaveCount(5);
    }

    [Fact]
    public void FindConnectedGroups_3つ以下のグループは無視する()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(0, 12, Cell.Filled(PuyoColor.Yellow))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Yellow))
            .SetCell(2, 12, Cell.Filled(PuyoColor.Yellow));

        // Act
        var groups = board.FindConnectedGroups();

        // Assert
        groups.Should().BeEmpty();
    }

    [Fact]
    public void ClearPuyos_指定された位置のぷよを消去できる()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(0, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(3, 12, Cell.Filled(PuyoColor.Red));

        // Act
        var positions = new List<(int X, int Y)> { (0, 12), (1, 12), (2, 12), (3, 12) };
        var newBoard = board.ClearPuyos(positions);

        // Assert
        newBoard.GetCell(0, 12).Should().Be(Cell.Empty);
        newBoard.GetCell(1, 12).Should().Be(Cell.Empty);
        newBoard.GetCell(2, 12).Should().Be(Cell.Empty);
        newBoard.GetCell(3, 12).Should().Be(Cell.Empty);
    }

    [Fact]
    public void ApplyGravity_浮いているぷよが落下する()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(2, 8, Cell.Filled(PuyoColor.Green))   // 浮いているぷよ
            .SetCell(2, 12, Cell.Filled(PuyoColor.Red));    // 下にあるぷよ

        // Act
        var newBoard = board.ApplyGravity();

        // Assert
        newBoard.GetCell(2, 8).Should().Be(Cell.Empty);
        newBoard.GetCell(2, 11).Should().Be(Cell.Filled(PuyoColor.Green));  // 落ちた
        newBoard.GetCell(2, 12).Should().Be(Cell.Filled(PuyoColor.Red));
    }

    [Fact]
    public void ApplyGravity_複数のぷよが落下する()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(1, 5, Cell.Filled(PuyoColor.Blue))
            .SetCell(1, 6, Cell.Filled(PuyoColor.Yellow))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Red));

        // Act
        var newBoard = board.ApplyGravity();

        // Assert
        newBoard.GetCell(1, 5).Should().Be(Cell.Empty);
        newBoard.GetCell(1, 6).Should().Be(Cell.Empty);
        newBoard.GetCell(1, 10).Should().Be(Cell.Filled(PuyoColor.Blue));
        newBoard.GetCell(1, 11).Should().Be(Cell.Filled(PuyoColor.Yellow));
        newBoard.GetCell(1, 12).Should().Be(Cell.Filled(PuyoColor.Red));
    }

    [Fact]
    public void 消去と重力の統合_消去後に上のぷよが落下する()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(0, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(3, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 10, Cell.Filled(PuyoColor.Blue));

        // Act - 消去して重力を適用
        var groups = board.FindConnectedGroups();
        var positions = groups.SelectMany(g => g).ToList();
        var clearedBoard = board.ClearPuyos(positions);
        var finalBoard = clearedBoard.ApplyGravity();

        // Assert
        finalBoard.GetCell(0, 12).Should().Be(Cell.Empty);
        finalBoard.GetCell(1, 12).Should().Be(Cell.Empty);
        finalBoard.GetCell(2, 12).Should().Be(Cell.Filled(PuyoColor.Blue));  // 青が落ちてきた
        finalBoard.GetCell(3, 12).Should().Be(Cell.Empty);
    }

    [Fact]
    public void ClearAndApplyGravityRepeatedly_消去対象がない場合は盤面が変わらない()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(0, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Blue));

        // Act
        var result = board.ClearAndApplyGravityRepeatedly();

        // Assert - 消去対象がないため、盤面は変わらない
        result.GetCell(0, 12).Should().Be(Cell.Filled(PuyoColor.Red));
        result.GetCell(1, 12).Should().Be(Cell.Filled(PuyoColor.Blue));
    }

    [Fact]
    public void ClearAndApplyGravityRepeatedly_2連鎖が正常に動作する()
    {
        // Arrange - 1連鎖目で赤が消え、2連鎖目で青が消えるパターン
        var board = Board.Create(6, 13);
        board = board
            // 赤ぷよ（2x2の正方形）
            .SetCell(1, 11, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 11, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 12, Cell.Filled(PuyoColor.Red))
            // 青ぷよ（縦に3つ + 横に1つ）
            .SetCell(3, 11, Cell.Filled(PuyoColor.Blue))
            .SetCell(2, 7, Cell.Filled(PuyoColor.Blue))
            .SetCell(2, 8, Cell.Filled(PuyoColor.Blue))
            .SetCell(2, 9, Cell.Filled(PuyoColor.Blue));

        // Act
        var result = board.ClearAndApplyGravityRepeatedly();

        // Assert - すべてのぷよが消えている（2連鎖が発生）
        for (int y = 0; y < 13; y++)
        {
            for (int x = 0; x < 6; x++)
            {
                result.GetCell(x, y).Should().Be(Cell.Empty);
            }
        }
    }

    [Fact]
    public void ClearAndApplyGravityRepeatedly_3連鎖が正常に動作する()
    {
        // Arrange - 3連鎖が発生するパターン
        var board = Board.Create(6, 13);
        board = board
            // 1連鎖目: 赤ぷよ（下部）
            .SetCell(0, 11, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 11, Cell.Filled(PuyoColor.Red))
            .SetCell(0, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Red))
            // 2連鎖目: 青ぷよ（中部）
            .SetCell(0, 6, Cell.Filled(PuyoColor.Blue))
            .SetCell(0, 7, Cell.Filled(PuyoColor.Blue))
            .SetCell(0, 8, Cell.Filled(PuyoColor.Blue))
            .SetCell(1, 8, Cell.Filled(PuyoColor.Blue))
            // 3連鎖目: 緑ぷよ（上部）
            .SetCell(0, 2, Cell.Filled(PuyoColor.Green))
            .SetCell(0, 3, Cell.Filled(PuyoColor.Green))
            .SetCell(0, 4, Cell.Filled(PuyoColor.Green))
            .SetCell(1, 4, Cell.Filled(PuyoColor.Green));

        // Act
        var result = board.ClearAndApplyGravityRepeatedly();

        // Assert - すべてのぷよが消えている（3連鎖が発生）
        for (int y = 0; y < 13; y++)
        {
            for (int x = 0; x < 6; x++)
            {
                result.GetCell(x, y).Should().Be(Cell.Empty);
            }
        }
    }

    [Fact]
    public void ClearAndApplyGravityRepeatedly_消去なしでも浮いたぷよが落下する()
    {
        // Arrange - 消去対象はないが、宙に浮いたぷよがある
        var board = Board.Create(6, 13);
        board = board
            .SetCell(2, 5, Cell.Filled(PuyoColor.Red))   // 浮いているぷよ
            .SetCell(2, 12, Cell.Filled(PuyoColor.Blue));  // 下にあるぷよ

        // Act
        var result = board.ClearAndApplyGravityRepeatedly();

        // Assert - 浮いていたぷよが落下している
        result.GetCell(2, 5).Should().Be(Cell.Empty);
        result.GetCell(2, 11).Should().Be(Cell.Filled(PuyoColor.Red));  // 落ちた位置
        result.GetCell(2, 12).Should().Be(Cell.Filled(PuyoColor.Blue));
    }

    [Fact]
    public void CheckZenkeshi_盤面が空の場合は全消しと判定される()
    {
        // Arrange
        var board = Board.Create(6, 13);

        // Act
        var isZenkeshi = board.CheckZenkeshi();

        // Assert
        isZenkeshi.Should().BeTrue();
    }

    [Fact]
    public void CheckZenkeshi_盤面にぷよが残っている場合は全消しと判定されない()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board.SetCell(3, 12, Cell.Filled(PuyoColor.Blue));

        // Act
        var isZenkeshi = board.CheckZenkeshi();

        // Assert
        isZenkeshi.Should().BeFalse();
    }

    [Fact]
    public void ClearAndApplyGravityRepeatedlyWithZenkeshi_全消しの場合はTrueを返す()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(1, 11, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 11, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 12, Cell.Filled(PuyoColor.Red));

        // Act
        var (finalBoard, isZenkeshi) = board.ClearAndApplyGravityRepeatedlyWithZenkeshi();

        // Assert
        isZenkeshi.Should().BeTrue();

        // すべてのセルが空であることを確認
        for (int y = 0; y < 13; y++)
        {
            for (int x = 0; x < 6; x++)
            {
                finalBoard.GetCell(x, y).Should().Be(Cell.Empty);
            }
        }
    }

    [Fact]
    public void ClearAndApplyGravityRepeatedlyWithZenkeshi_全消しでない場合はFalseを返す()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(0, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Blue));

        // Act
        var (finalBoard, isZenkeshi) = board.ClearAndApplyGravityRepeatedlyWithZenkeshi();

        // Assert
        isZenkeshi.Should().BeFalse();

        // ぷよが残っていることを確認
        finalBoard.GetCell(0, 12).Should().Be(Cell.Filled(PuyoColor.Red));
        finalBoard.GetCell(1, 12).Should().Be(Cell.Filled(PuyoColor.Blue));
    }

    [Fact]
    public void ClearAndApplyGravityRepeatedlyWithChainInfo_1連鎖で4個消去した場合は消去数リストに4を返す()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(1, 11, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 11, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(2, 12, Cell.Filled(PuyoColor.Red));

        // Act
        var (finalBoard, clearedCounts, isZenkeshi) = board.ClearAndApplyGravityRepeatedlyWithChainInfo();

        // Assert
        clearedCounts.Should().HaveCount(1);
        clearedCounts[0].Should().Be(4);
        isZenkeshi.Should().BeTrue();
    }

    [Fact]
    public void ClearAndApplyGravityRepeatedlyWithChainInfo_2連鎖の場合は消去数リストに各連鎖の消去数を返す()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            // 1連鎖目: 赤4個（縦一列、x=1）
            .SetCell(1, 9, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 10, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 11, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Red))
            // 2連鎖目: 青3個（横、y=12、x=2-4）+ 青1個（赤の上、x=1, y=8）
            .SetCell(2, 12, Cell.Filled(PuyoColor.Blue))
            .SetCell(3, 12, Cell.Filled(PuyoColor.Blue))
            .SetCell(4, 12, Cell.Filled(PuyoColor.Blue))
            // この青が赤の上、赤が消えると落下して横3個と繋がる
            .SetCell(1, 8, Cell.Filled(PuyoColor.Blue));

        // Act
        var (finalBoard, clearedCounts, isZenkeshi) = board.ClearAndApplyGravityRepeatedlyWithChainInfo();

        // Assert
        clearedCounts.Should().HaveCount(2);
        clearedCounts[0].Should().Be(4); // 1連鎖目: 赤4個
        clearedCounts[1].Should().Be(4); // 2連鎖目: 青4個
        isZenkeshi.Should().BeTrue();
    }

    [Fact]
    public void ClearAndApplyGravityRepeatedlyWithChainInfo_消去なしの場合は空のリストを返す()
    {
        // Arrange
        var board = Board.Create(6, 13);
        board = board
            .SetCell(0, 12, Cell.Filled(PuyoColor.Red))
            .SetCell(1, 12, Cell.Filled(PuyoColor.Blue));

        // Act
        var (finalBoard, clearedCounts, isZenkeshi) = board.ClearAndApplyGravityRepeatedlyWithChainInfo();

        // Assert
        clearedCounts.Should().BeEmpty();
        isZenkeshi.Should().BeFalse();
    }
}
